librad/src/paths.rs

```diff
@@ -7,38 +8,104 @@ use std::{
 use directories::ProjectDirs;
 
 #[derive(Clone)]
-pub struct Paths(ProjectDirs);
+pub struct Paths {
+    keys_dir: PathBuf,
+    projects_dir: PathBuf,
+    profiles_dir: PathBuf,
+}
 
 impl Paths {
     pub fn new() -> Result<Self, io::Error> {
-        let proj = ProjectDirs::from("xyz", "radicle", "radicle")
-            .expect("Unable to determine application directories");
-        Paths(proj).init()
+        let proj = ProjectDirs::from("xyz", "radicle", "radicle").ok_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::NotFound,
+                "Couldn't determine application directories.",
+            )
+        })?;
+
+        let config_dir = proj.config_dir();
+        let data_dir = proj.data_dir();
+
+        Self {
+            keys_dir: config_dir.join("keys"),
+            projects_dir: data_dir.join("projects"),
+            profiles_dir: config_dir.join("profiles"),
+        }
+        .init()
     }
 
     // Don't use system paths, but the supplied directory as a root.
     //
     // For testing, you know.
     pub fn from_root(root: &Path) -> Result<Self, io::Error> {
-        Paths(ProjectDirs::from_path(root.to_path_buf()).unwrap()).init()
+        Self {
+            keys_dir: root.join("keys"),
+            projects_dir: root.join("projects"),
+            profiles_dir: root.join("profiles"),
+        }
+        .init()
     }
 
     pub fn keys_dir(&self) -> PathBuf {
-        self.0.config_dir().join("keys")
+        self.keys_dir.clone()
     }
 
     pub fn projects_dir(&self) -> PathBuf {
-        self.0.data_dir().join("projects")
+        self.projects_dir.clone()
     }
 
     pub fn profiles_dir(&self) -> PathBuf {
-        self.0.config_dir().join("profiles")
+        self.profiles_dir.clone()
+    }
+
+    pub fn all(&self) -> HashMap<&str, PathBuf> {
+        // Nb. this pattern match is here to keep the map consistent with the
+        // struct fields
+        match self {
+            Self {
+                keys_dir,
+                projects_dir,
+                profiles_dir,
+            } => [
+                ("keys_dir", keys_dir.clone()),
+                ("projects_dir", projects_dir.clone()),
+                ("profiles_dir", profiles_dir.clone()),
+            ]
+            .iter()
```

Comment:
I think `iter().cloned()` is the same as doing `into_iter()` but not 100% positive on that

Reply:
Doesn't seem so, `into_iter()` yields `Iterator<Item=&(&str, &Path)>`, but I need `Iterator<Item=(&str, &Path)>` (i.e. the tuple needs to be cloned, not the values inside).

Reply:
Ah ok, my bad